### DISCUSSION TITLE
Typing: Correct way to specify MutableDict.as_mutable(JSONB....

### DISCUSSION THREAD

**Discussion by bigmikef** [sqlalchemy/sqlalchemy#12046] [Usage Questions]
Please forgive me as this is my first post in this forum.

Please consider the following code in python3.12 with pip installed of SQLAlchemy-2.0.36 and mypy-1.13.0
```python
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.ext.mutable import MutableDict
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class SimpleEntity(DeclarativeBase):
    __tablename__ = "simple"

    no_paren: Mapped[MutableDict] = mapped_column(
        "no_paren", MutableDict.as_mutable(JSONB)  # produces the mypy error
    )

    paren: Mapped[MutableDict] = mapped_column("paren", MutableDict.as_mutable(JSONB()))
```

Running mypy on the above file yields an error for the non_paren `JSONB` but the `JSONB()` does not error.
```
$ mypy simple_entity.py
simple_entity.py:10: error: Argument 1 to "as_mutable" of "Mutable" has incompatible type "type[JSONB]"; expected "TypeEngine[Any]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

My questions are:
- Which of the above two are correct?
- Are they both valid, and I should `#type :ignore` the one without the parentheses?

**Comment by bigmikef** [comment]
Some links to other discussions on the issue
- Suggestion to use with_variant in similar but not matching issue - https://github.com/sqlalchemy/sqlalchemy/discussions/9530
- Another discussion to use JSONB() - https://github.com/sqlalchemy/sqlalchemy/discussions/11276#discussioncomment-9130637

**Comment by bigmikef** [comment]
Where I'm confused is why it works with instantiating the JSONB object, but the raw type does not work.

**Comment by CaselIT** [comment]
Hi,

It's just the typing that's wrong, since it should accept both. Feel free to create a PR to fix it

**Reply by CaselIT** [reply]
Created https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5554. Will merge once it passes

**Reply by bigmikef** [reply]
Thank you so much
